driver/native/native.go

```diff
@@ -142,33 +152,58 @@ func (r *parseResponse) UnmarshalJSON(data []byte) error {
 }
 
 func (d *Driver) writeRequest(ctx context.Context, req *parseRequest) error {
-	deadline, _ := ctx.Deadline()
-
 	sp, _ := opentracing.StartSpanFromContext(ctx, "bblfsh.native.Parse.encodeReq")
 	defer sp.Finish()
 
-	if !deadline.IsZero() {
-		d.stdin.SetWriteDeadline(deadline)
-		defer d.stdin.SetWriteDeadline(time.Time{})
-	}
-
 	return d.enc.Encode(req)
 }
 
-func (d *Driver) readResponse(ctx context.Context) (*parseResponse, error) {
-	deadline, _ := ctx.Deadline()
+type timeoutError interface {
+	Timeout() bool
+}
 
-	sp, _ := opentracing.StartSpanFromContext(ctx, "bblfsh.native.Parse.decodeResp")
+func (d *Driver) broken(err error) {
+	d.state = stateBroken
+	d.lastErr = err
+	_ = d.Close()
+}
+
+func (d *Driver) skipResponse(ctx context.Context) error {
+	if d.state != stateTimeout { // safeguard
+		panic(fmt.Errorf("unexpected state: %v", d.state))
+	}
+	sp, _ := opentracing.StartSpanFromContext(ctx, "bblfsh.native.Parse.skipResp")
 	defer sp.Finish()
 
-	if !deadline.IsZero() {
-		d.stdout.SetReadDeadline(deadline)
-		defer d.stdout.SetReadDeadline(time.Time{})
+	// TODO: ideally we should just skip it without decoding
+	var r struct{}
```

Comment:
You can avoid most of the work of decoding by passing `var r json.RawMessage`. It will still have to parse the message, but will not construct memory objects for it apart from the byte array.

Reply:
I actually thought the same and then stopped myself thinking that encoder might not be JSON. But for now it is, so I'll change it.